### PR TITLE
issue/1333 Added error abstraction

### PR DESCRIPTION
**Pull request by oliverfoster**
[#1333](https://github.com/adaptlearning/adapt_framework/issues/1333)
[#1238](https://github.com/adaptlearning/adapt_framework/issues/1238)

### Added
* Error objects with codes and error data
* Default messages and translatable overrides
* Used an `_isCancellable: false` notify to display `CLIENT_COULD_NOT_CONNECT` errors

image

### CODE REVIEW

File: js/scorm/wrapper.js [hunk 1]
```diff
@@ -465,12 +478,33 @@ define([
...
+        Adapt.once('contentObjectView:ready', this.handleError.bind(this, error));
+        return;
+      }
+
+      const config = Adapt.course.get('_spoor');
+      const errorMessages = Object.assign({}, ScormError.defaultMessages, config && config._errorMessages);
+      const message = Handlebars.compile(errorMessages[error.name])(error.data);
+
+      switch (error.name) {
+        case CLIENT_COULD_NOT_CONNECT:
```
Comment (oliverfoster) [inline js/scorm/wrapper.js:493]: @moloko Not sure if you want to handle any other errors this way, or if you want to be able to configure?

**Comment by oliverfoster** [comment]
There are also a series of other logs which may need to be translatable? like: https://github.com/adaptlearning/adapt-contrib-spoor/blob/master/js/scorm/wrapper.js#L75

File: js/scorm/wrapper.js (hunk 1 above)
Comment (moloko) [reply js/scorm/wrapper.js:493]: TBH 'could not connect' is the big one... the other one that *might* be useful is [this one](https://github.com/adaptlearning/adapt-contrib-spoor/blob/68eb444bac3926b69da56cb98061b142d3664fe6/js/scorm/wrapper.js#L254)

Comment (moloko) [reply js/scorm/wrapper.js:493]: ... which I notice now you've already handled!

Comment (oliverfoster) [reply js/scorm/wrapper.js:493]: 'Course is not connected to the LMS' doesn't display as a notify, only 'Could not connect' is notified atm.

Comment (moloko) [reply js/scorm/wrapper.js:493]: that's absolutely fine

Comment (oliverfoster) [reply js/scorm/wrapper.js:493]: I figured I'd do all the ones that end up with a user popup whilst I was there. Not sure how useful that is?

File: js/scorm/wrapper.js [hunk 2]
```diff
@@ -465,12 +478,33 @@ define([
...
+            title: errorMessages['title'],
+            body: message
+          });
+          return;
+      }
+
+      this.logger.error(message);
 
-      if (!this.suppressErrors && (!this.logOutputWin || this.logOutputWin.closed) && confirm(`An error has occured:\n\n${_msg}\n\nPress 'OK' to view debug information to send to technical support.`)) {
+      if (!this.suppressErrors && (!this.logOutputWin || this.logOutputWin.closed) && confirm(`An error has occured:\n\n${message}\n\nPress 'OK' to view debug information to send to technical support.`)) {
```
Comment (oliverfoster) [inline js/scorm/wrapper.js:504]: The classic popup pre/post-amble probably needs adding somewhere/how

File: js/scorm/wrapper.js (hunk 1 above)
Comment (moloko) [reply js/scorm/wrapper.js:493]: I think it's only right and proper that Adapt, claiming to have multilanguage support, should be able to display error messages in the learner's choice of language!

**Review by moloko — APPROVED** [review]

**Review by lc-alexanderbenesch — APPROVED** [review]
👍

**Pull request merged**